design-documents/remote-invoker.md

```diff
@@ -0,0 +1,631 @@
+multiple _Remote Code Invoker_ implementations based on existing Web APIs, Message Queue RPC API or any other.
+### How are remote services are going to look like with Remote Code Invoker
+\<Module name\>Proxy module will contain \<Module name\>API module's implementations that would look like this:
+```
+class ServiceProxy implements \Magento\ModuleNameApi\ServiceInterface
+{
+
+....
+
+    public function doStuff(\Magento\ModuleNameApi\Data\DTOInterface $dto): \Magento\ModuleNameApi\Data\ResultDTOInterface
+    {
+        $remoteResult = null;
+        $promise = $this->invoker->invoke(
+            'catalog',
+            \Magento\ModuleNameApi\ServiceInterface::class,
+            'doStuff',
+            [$dto]
+        );
+        $promise->then(function (\Magento\ModuleNameApi\Data\ResultDTOInterface $result) use (&$remoteResult) {
+            $remoteResult = $result;
+        });
+        $promise->wait();
+        
+        return $remoteResult;
+    }
+
+....
+
+}
+``` 
+ 
+di.xml in \<Module name\>Proxy module will specify that ServiceProxy
+is the preference for \Magento\ModuleNameApi\ServiceInterface.
+
+So when a local service uses \<Module name\>API's service it will not be aware whether the service is actually
+remote and it won't affect how we write code.
+Example:
+```
+class LocalServiceA implements ServiceAInterface
+{
+
+    /**
+     * @var \Magento\BModuleApi\ServiceBInterface
+     */
+    private $serviceB;
+
+....
+
+    public function doAStuff(): void
+    {
+        ....
+
+        $stuff = $this->serviceB->doBStuff($someVar);
+        
+        ....
+    }
+
+....
+
+}
+```
+
+Depending on whether _BModule_ or _BModuleProxy_ is deployed on the local node _ServiceBInterface_ will either
+work locally or remotely.
+### Why not just use cURL and existing Web APIs
+We need a mechanism, an abstraction around an actual network interactions that would provide a retries mechanism,
+a standard to pass local call's arguments, a request signing mechanism, authentication mechanism which would allow
+to perform authentication only once on the node that accepts user requests and not on each node, data consistency
+mechanism etc.
+_([see explanation](#need-for-an-abstraction-around-network-calls))_.
+### Functional requirements
+* Invoker's code is a part of Magento\Framework namespace
+* Invoker can call classes and method that were specifically enabled for remote calls via a config
+* Invoker provides a way for services to written as usual, not knowing whether other modules' services are remote or not
+* Multiple remote sources can be configured and assigned unique IDs
+* Each HTTP request to remote sources is signed using unique source's key, timestamp and request's attributes
+* Each remote call must be given unique ID to be used by retries mechanism
+* Timeout limit and retries count can be configured globally and for each call
```

Comment:
I think service mesh should handle these things.

Reply:
service mash is a separate system which is acting as a proxy between inter-service communications and, yes, provides retries and data consistency. I haven't found a mention of such system in Anton's docs and nor do we need such system - they only make sense when different services use different languages/frameworks etc. And we are going to use PHP + Magento on all of our services.

Reply:
If what you were meaning to say is that we need a separate entity in our codebase to handle retries, data consistency etc - this document is in fact describing such entity (invoker = service mesh)

Reply:
Service mesh can solve many problems related to deployment and communication for us: service registration, routing, load balancing, circuit breaking, retries, statistics and monitoring, etc. Retries logic is complicated and I would suggest not to implement it ourselves.